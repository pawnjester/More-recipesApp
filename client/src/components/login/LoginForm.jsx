import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import swal from 'sweetalert';
import history from '../../utils/history';
import TextFieldGroupComponent from '../common/TextFieldGroup';
import validateInput from '../validations/loginValidations';
import { login } from '../../actions/loginActions';

/**
 *
 * @class LoginForm
 *
 * @extends {React.Component}
 */
export class LoginForm extends React.Component {
  /**
   * Creates an instance of LoginForm.
   *
   * @param {any} props
   *
   * @memberof LoginForm
   */
  constructor(props) {
    super(props);
    this.state = {
      identifier: '',
      password: '',
      errors: {},
      isLoading: false,
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }
  /**
 * @description submit action
 *
 * @param {any} event
 *
 * @memberof LoginForm
 *
 * @returns {void}
 */
  onSubmit(event) {
    event.preventDefault();

    if (this.isValid()) {
      this.setState({ errors: {}, isLoading: true });
      this.props.login(this.state).then(() => {
        swal({
          title: 'welcome',
          text: this.state.username,
          icon: 'success'
        });
        history.push('/recipes');
      })
        .catch((err) => {
          swal({
            title: 'Oops!',
            text: `Sorry ${err.response.data.error}`,
            icon: 'warning'
          });
          this.setState({
            isLoading: false
          });
        });
    }
  }
  /**
 *
 * @param {any} event
 *
 * @memberof LoginForm
 *
 * @returns {void}
 */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }
  /**
 *
 * @memberof LoginForm
 *
 * @returns {void}
 */
  isValid() {
    const { errors, isValid } = validateInput(this.state);

    if (!isValid) {
      this.setState({ errors });
    }

    return isValid;
  }
  /**
 *@description renders jsx element
 *
 * @memberof LoginForm
 *
 * @returns {void}
 */
  render() {
    const { errors } = this.state;
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {errors.message && <div className="alert alert-danger">{errors.message}</div>}
          <TextFieldGroupComponent
            error={errors.identifier || errors.errorname}
            onChange={this.onChange}
            value={this.state.identifier}
            field="identifier"
            placeholder="enter your username or email"
            autofocus
            required
          />

          <TextFieldGroupComponent
            error={errors.password}
            onChange={this.onChange}
            value={this.state.password}
            field="password"
            placeholder="password"
            type="password"
          />
          <input
            disabled={this.state.isLoading}
            type="submit"
            value="Submit"
            className="btn btn-outline-danger btn-block text-dark login-submit"
          />
          <div>
            <Link to="/reset_password" className="forgot mt-5">
            Forgot Password?
            </Link><br />
            <Link to="/" className="mt-5">Sign Up</Link>
          </div>
        </form>
      </div>
    );
  }
}

LoginForm.propTypes = {
  login: PropTypes.func.isRequired,
};
export default connect(null, { login })(LoginForm);
