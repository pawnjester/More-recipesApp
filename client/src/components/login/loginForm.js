import React from 'react';
import TextFieldGroup from '../common/TextFieldGroup';
import validateInput from './validations';
import  {connect} from 'react-redux';
import {login} from '../../actions/loginActions';
import PropTypes from 'prop-types';
import {Redirect} from  'react-router-dom'


class LoginForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      username : '',
      password: '',
      errors: {},
      isLoading: false,
      redirect: false
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  isValid() {
    const { errors, isValid } = validateInput(this.state);

    if(!isValid) {
      this.setState({errors})
    }

    return isValid;

  }

  onSubmit(e) {
    e.preventDefault();

    if(this.isValid()){
      this.setState({errors: {}, isLoading: true});
      this.props.login(this.state).then(
      (res) => {
        this.setState({redirect : true})},
        (err) => this.setState({ errors: err.response.data, isLoading: false }),
      );

    }
  }

  onChange(e) {
        this.setState({ [e.target.name]: e.target.value})
  }

  render() {
  const {errors, redirect} = this.state;
  if (redirect) {
       return <Redirect to='/'/>;
     }
    return (
    <form onSubmit = {this.onSubmit}>
      {errors.message && <div className ="alert alert-danger">{errors.message}</div>}
      <TextFieldGroup
                    error = {errors.username || errors.errorname}
                    onChange = {this.onChange}
                    value = {this.state.username}
                    field = "username"
                    placeholder = "username" 
                    autofocus required
                  />
      
      <TextFieldGroup
                    error = {errors.password}
                    onChange = {this.onChange}
                    value = {this.state.password}
                    field = "password"
                    placeholder = "password"
                    
                  />
                  <input 
                    disabled = {this.state.isLoading}
                    type="submit"
                    value ="Submit"
                    className="btn btn-outline-danger btn-block text-dark" />
    </form>

    )
  }
}

LoginForm.PropTypes ={
  login: PropTypes.func.isRequired
}
export default connect(null, { login})(LoginForm)
