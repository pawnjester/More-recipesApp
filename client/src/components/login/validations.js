import Validator from 'validator';
import isEmpty from 'lodash/isEmpty';
/**
 *
 *
 * @export
 * @param {any} data
 * @returns
 */
export default function validatorInput(data) {
  const errors = {};

  if (Validator.isEmpty(data.username)) {
    errors.username = 'This field is required';
  } else if (data.username.length < 6) {
    errors.username = 'Username is too short, Must be min. of 6';
  }

  if (Validator.isEmpty(data.password)) {
    errors.password = 'This field is required';
  } else if (data.password.length < 6) {
    errors.password = 'Password is too short, Must be min. of 6 ';
  }


  return {
    errors,
    isValid: isEmpty(errors),
  };
}
