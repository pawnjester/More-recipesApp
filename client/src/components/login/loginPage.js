import React from 'react';
import '../../styles/signup.scss';
import LoginForm from './loginForm';



class LoginPage extends React.Component {
  render() {
    return (
      <header id ="home-section">
      <div className= "dark-overlay">
      <div className= "home-inner">
        <div className ="container">
          <div className= "row justify-content-center">
            <div className='col-lg-6'>
              <div className= "card bg-white text-center card-form">
                <div className="card-body">
                  <h3>Sign In</h3>
                  <p>Welcome to More-Recipes, Please log in</p>
                  <LoginForm />
                </div>
              </div>
            </div>
            
          </div>
          
        </div>
        
      </div>
      </div>
      
    </header>
    )
  }
}

export default LoginPage;