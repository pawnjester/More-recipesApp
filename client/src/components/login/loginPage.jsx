import React from 'react';
import LoginForm from './loginForm';
import { connect } from 'react-redux';
import {login} from '../../actions/loginActions';
import Proptypes from 'prop-types';



class LoginPage extends React.Component {
  render() {
    return (
      
      <div className= "home-inner">
        <div className ="container">
          <div className= "row justify-content-center">
            <div className='col-lg-6'>
              <div className= "card bg-white text-center card-form">
                <div className="card-body">
                  <h3>Sign In</h3>
                  <p>Welcome to More-Recipes, Please log in</p>
                  <LoginForm login = {login}/>
                </div>
              </div>
            </div>
            
          </div>
          
        </div>
        
      </div>
      
    )
  }
}

LoginPage.proptypes ={
  login: Proptypes.func.isRequired
}

export default connect(null, { login})(LoginPage)