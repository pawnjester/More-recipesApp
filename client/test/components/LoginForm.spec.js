import React from 'React';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect';
import { LoginForm } from '../../src/components/login/LoginForm';
import store from '../../src/store/store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import sinon from 'sinon';


configure({ adapter: new Adapter() });

const props = {
  login: jest.fn(() => Promise.resolve())
};

function setup() {
  const shallowWrapper = shallow(<LoginForm {...props} />);
  return {
    shallowWrapper,
  };
};

describe('Login Form Component', () => {
  it('should render correctly', () => {
    const { shallowWrapper } = setup();
    expect(shallowWrapper).toMatchSnapshot();
  });
});

describe('onSubmit() should', () => {
  it('be called when form is submitted', () => {
    const { shallowWrapper } = setup();
    const event = {
      preventDefault: jest.fn()
    };
    shallowWrapper.setState({
      identifier: 'ebuke@example.com',
      password: 'chalrsss',
    });
    shallowWrapper.instance().onSubmit(event);
    expect(props.login.mock.calls.length).toEqual(1);
});
});

describe('onChange() should', () => {
  it('be called when form input changes', () => {
    const { shallowWrapper } = setup();
    const event = {
      target: { name: 'identifier', value: 'phil' }
    };
    shallowWrapper.instance().onChange(event);
  });
});
