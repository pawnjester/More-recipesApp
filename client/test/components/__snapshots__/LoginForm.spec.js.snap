// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form Component should render correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <TextFieldGroup
      autofocus={true}
      error=""
      field="identifier"
      onChange={[Function]}
      placeholder="enter your username or email"
      required={true}
      type="text"
      value=""
    />
    <TextFieldGroup
      error=""
      field="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <input
      className="btn btn-outline-danger btn-block text-dark login-submit"
      disabled={false}
      type="submit"
      value="Submit"
    />
    <div>
      <Link
        className="forgot mt-5"
        replace={false}
        to="/reset_password"
      >
        Forgot Password?
      </Link>
      <br />
      <Link
        className="mt-5"
        replace={false}
        to="/"
      >
        Sign Up
      </Link>
    </div>
  </form>
</div>
`;
